### PR TITLE
Adding support for Scalar Crypto Extension (Bitmanip instructions for Cryptography, Zbkb)

### PR DESCRIPTION
**Pull request by munailwaqar**
### **Introduction**
This PR adds support for **Zbkb extension** in the CVA6 core. It also adds the documentation for this extension. These changes have been tested with self-written single instruction tests and with the riscv-arch-tests. This PR will be followed by other PRs that will add complete support for the **Zkn - NIST Algorithm Suite** extension.

### **Implementation**
Zbkb Extension:
Added support for the Zbkb instruction set. It essentially expands the Zbb extension with additional instructions useful in cryptography. These instructions are pack, packh, packw, brev8, unzip and zip.

### **Modifications**
1. A new bit **ZKN** was added. The complete Zkn extension will be added under this bit for ease of use. This configuration will also require the RVB (bitmanip) bit to be set.
2. Updated the ALU and decoder to recognize and handle Zbkb instructions.

### **Documentation and Reference**
The official RISC-V Cryptography Extensions Volume I was followed to ensure alignment with ratification. The relevant documentation for the Zbkb instruction was also added.

### **Verification**
**Assembly Tests:**
The instructions were tested and verified with the K module of both 32 bit and 64 bit versions of the riscv-arch-tests to ensure proper functionality. These tests check for ISA compliance, edge cases and use assertions to ensure expected behavior. The tests include:
1. pack-01.S 
2. packh-01.S
3. packw-01.S
4. brev8-01.S
5. unzip-01.S
6. zip-01.S

### CODE REVIEW

**Comment by JeanRochCoulon** [comment]
Great!    To solve the Verible format issues, I suggest to use the command reported in CONTRIBUTING.md.
Do not forget to sign the ECA agreement to bcome cva6 contributor.

**Comment by JeanRochCoulon** [comment]
First comments:
- cv32a60x and cv32a65x do not need the new extension. I suggest to apply the extension the configuration used for FPGA: cv32a6_imac_sv32 and disable the extension for 60x and 65x.
- In config_pkg.sv files, better is to not define localparam. You can follow the 65x_config_pkg.sv example.
Cheers

**Comment by fatimasaleem** [comment]
> First comments:
> 
> * cv32a60x and cv32a65x do not need the new extension. I suggest to apply the extension the configuration used for FPGA: cv32a6_imac_sv32 and disable the extension for 60x and 65x.
> * In config_pkg.sv files, better is to not define localparam. You can follow the 65x_config_pkg.sv example.
>   Cheers

Thanks @JeanRochCoulon . Sure, we will disable it for `cv32a60x` and `cv32a65x` . We only wanted to run the CI with this extension enabled. Since now the CI is almost green, we will disable it for these two.

File: verif/sim/cva6.py [hunk 1]
```diff
@@ -879,14 +879,14 @@ def load_config(args, cwd):
       args.isa  = "rv64gch_zba_zbb_zbs_zbc"
     elif base in ("cv64a6_imafdc_sv39", "cv64a6_imafdc_sv39_hpdcache", "cv64a6_imafdc_sv39_wb"):
       args.mabi = "lp64d"
-      args.isa  = "rv64gc_zba_zbb_zbs_zbc"
```
Comment (JeanRochCoulon) [inline verif/sim/cva6.py:882]: cv64a6_imafdc_sv39_wb does not support zbkb, this has to be fixed

File: verif/sim/cva6.py [hunk 2]
```diff
@@ -879,14 +879,14 @@ def load_config(args, cwd):
...
+      args.isa  = "rv64gc_zba_zbb_zbs_zbc_zbkb"
     elif base == "cv32a60x":
       args.mabi = "ilp32"
-      args.isa  = "rv32imc_zba_zbb_zbs_zbc"
```
Comment (JeanRochCoulon) [inline verif/sim/cva6.py:885]: 60x and 65x do not support zbkb

File: verif/sim/cva6.py (hunk 1 above)
Comment (fatimasaleem) [reply verif/sim/cva6.py:882]: we want to enable this extension for the basic 64-bit config as well, which is `cv64a6_imafdc_sv39` 

@munailwaqar lets only enable it for cv64a6_imafdc_sv39.

**Comment by munailwaqar** [comment]
Currently enabling Zbkb in 2 configs only: 
1. cv32a6_imac_sv32
2. cv64a6_imafdc_sv39

**Comment by fatimasaleem** [comment]
Hi @JeanRochCoulon, can you please approve this?

File: core/include/cv32a6_imac_sv32_config_pkg.sv [hunk 3]
```diff
@@ -24,7 +24,7 @@ package cva6_config_pkg;
   localparam CVA6ConfigZcmpExtEn = 0;
   localparam CVA6ConfigAExtEn = 1;
   localparam CVA6ConfigHExtEn = 0;  // always disabled
-  localparam CVA6ConfigBExtEn = 0;
+  localparam CVA6ConfigBExtEn = 1;
```
Comment (JeanRochCoulon) [inline core/include/cv32a6_imac_sv32_config_pkg.sv:27]: This localparam can be removed, because not used

**Comment by JeanRochCoulon** [comment]
Hello @munailwaqar 
Great to have identified riscv-arc-test tests to test the new extension. Can you add these test to the testlist of the 64bit configuration (cv64a6_imafdc_sv39) to insert them into CI and ease the maintenance.

**Comment by munailwaqar** [comment]
> Hello @munailwaqar Great to have identified riscv-arc-test tests to test the new extension. Can you add these test to the testlist of the 64bit configuration (cv64a6_imafdc_sv39) to insert them into CI and ease the maintenance.

Added the relevant arch tests to the testlist but I also had to enable the Zbkb extension in cv64a6_imafdc_sv39_hpdcache config as the CI runs checks for this config too using the same testlist.

**Comment by fatimasaleem** [comment]
Hi @JeanRochCoulon can you please review this again? Thanks.

**Comment by munailwaqar** [comment]
> Hello @munailwaqar zip and unzip tests are not included in testlist, is it intentional ?

@JeanRochCoulon Yes. zip and unzip instructions are only for 32bit architectures according to the documentation.

**Comment by JeanRochCoulon** [comment]
I cannot rebase because I have no right on your branch, can you ?

**Comment by munailwaqar** [comment]
Yes ill do it.

**Comment by JeanRochCoulon** [comment]
LGTM. I am ready to merge.
Can you rebase (again) then, taking back the resulting gate count, adjust the expected gate count ?
